homeassistant/components/powerwall/binary_sensor.py

```diff
@@ -0,0 +1,183 @@
+from .const import (
+    ATTR_GRID_CODE,
+    ATTR_NOMINAL_SYSTEM_POWER,
+    ATTR_REGION,
+    DOMAIN,
+    POWERWALL_API_GRID_STATUS,
+    POWERWALL_API_SITEMASTER,
+    POWERWALL_CONNECTED_KEY,
+    POWERWALL_COORDINATOR,
+    POWERWALL_GRID_CODE,
+    POWERWALL_GRID_ONLINE,
+    POWERWALL_IP_ADDRESS,
+    POWERWALL_NOMINAL_SYSTEM_POWER,
+    POWERWALL_REGION,
+    POWERWALL_RUNNING_KEY,
+    POWERWALL_SITE_INFO,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the August sensors."""
+    powerwall_data = hass.data[DOMAIN][config_entry.entry_id]
+
+    coordinator = powerwall_data[POWERWALL_COORDINATOR]
+    site_info = powerwall_data[POWERWALL_SITE_INFO]
+    ip_address = powerwall_data[POWERWALL_IP_ADDRESS]
+
+    entities = []
+    entities.append(PowerWallRunningSensor(coordinator, site_info, ip_address))
+    entities.append(PowerWallGridStatusSensor(coordinator, site_info, ip_address))
+    entities.append(PowerWallConnectedSensor(coordinator, site_info, ip_address))
+
+    async_add_entities(entities, True)
+
+
+class PowerWallBinarySensor(PowerWallEntity, BinarySensorDevice):
+    """Base class for powerwall binary sensors."""
+
+    def __init__(self, coordinator, site_info, ip_address):
+        """Initialize the sensor."""
+        super().__init__(coordinator, site_info, ip_address)
+        self._coordinator = coordinator
+
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        return self._coordinator.last_update_success
+
+    @property
+    def should_poll(self):
+        """Return False, updates are controlled via coordinator."""
+        return False
+
+    async def async_update(self):
+        """Update the entity.
+
+        Only used by the generic entity update service.
+        """
+        await self._coordinator.async_request_refresh()
+
+    async def async_added_to_hass(self):
+        """Subscribe to updates."""
+        self._coordinator.async_add_listener(self.async_write_ha_state)
+
+    async def async_will_remove_from_hass(self):
+        """Undo subscription."""
+        self._coordinator.async_remove_listener(self.async_write_ha_state)
+
+
+class PowerWallRunningSensor(PowerWallBinarySensor):
+    """Representation of an Powerwall running sensor."""
+
+    def __init__(self, coordinator, site_info, ip_address):
+        """Initialize the sensor."""
+        super().__init__(coordinator, site_info, ip_address)
+        self._coordinator = coordinator
+        self._site_info = site_info
+        self._ip_address = ip_address
+
+    @property
+    def name(self):
+        """Device Name."""
+        return "Powerwall Status"
+
+    @property
+    def device_class(self):
+        """Device Class."""
+        return DEVICE_CLASS_POWER
+
+    @property
+    def unique_id(self):
+        """Device Uniqueid."""
+        return f"{self._ip_address}_running"
```

Comment:
IP address is not allowed as unique id, as it may change. MAC address would be ok. Serial number is the best.

https://developers.home-assistant.io/docs/entity_registry_index#unique-id-requirements

Reply:
@MartinHjelmare Since tesla uses the serial number as a password they don't expose it via the api.  I could use
```
"_".join(site_info["utility"],site_info["grid_code"],site_info["nominal_system_power_kW"],site_info["nominal_system_energy_kWh"])
```
That should be unique unless they have two exactly the same powerwall systems with the exact same number of batteries connected to the exact same grid region on the same local network which Tesla would never install.

Reply:
Are you not allowed to install more than one powerwall per house?

Reply:
seems to work fine

Reply:
I'm not familiar with the four terms in the group. Please explain in more detail how that group makes the id unique and stable.

Reply:
utility is the utility that the power wall is connected to. This is static and cannot since the powerwall won't be physically relocated 

The grid code is the configuration of the incoming power

The nominal_system_power_kW is how much output the combination of all installed powerwalls on a site can output 

The nominal_system_energy_kWh is how much energy the device can store.

Since power walls are local and there is only supposed to be one gateway per site we should never see two systems but there may be a use case where your LAN is connected to multiple homes or you have a multi site vpn so I didn't want to use just the nominal_system_power_kW

Reply:
If an additional power wall is installed after the first one, will the nominal_system_power_kW change then?

Reply:
If you add a new powerwall to the system nominal_system_power_kW will go up.  Its all abstracted away by the gateway though so you'll just see the available power go up.  The gateway represents the combination of powerwalls as a single system

Reply:
So then the unique_id is not stable?

Reply:
> So then the unique_id is not stable?

I guess the question is that if you add a powerwall to the system is in a new system or is it the same system?  This is all very much theoretical and that will probably never happen though as adding power walls to the system requires permits, electrical work and would only likely happen if the user added additional solar panels as well.  

That being said should remove `nominal_system_power_kW` from the join?

Reply:
The id has to be both unique and stable per device. If the existing device isn't considered the same device after an additional power wall is installed, I guess this could work.

Reply:
I'll take the `nominal_system_power_kW` out in a followup

Reply:
https://github.com/home-assistant/core/pull/33021